server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -260,13 +300,19 @@ public DesiredBalance compute(
                 break;
             }
             if (i % 100 == 0) {
-                // TODO this warning should be time based, iteration count should be proportional to the number of shards
+                final int iterations = i;
+                final long currentTime = threadPool.relativeTimeInMillis();
+                final boolean shouldWarnByTime = nextWarningTime <= currentTime;
+                if (shouldWarnByTime) {
+                    nextWarningTime = currentTime + timeWarningInterval;
+                }
                 logger.log(
-                    i % 1000000 == 0 ? Level.INFO : Level.DEBUG,
-                    Strings.format(
-                        "Desired balance computation for [%d] is still not converged after [%d] iterations",
+                    i % iterationCountWarningInterval == 0 || shouldWarnByTime ? Level.INFO : Level.DEBUG,
+                    () -> Strings.format(
+                        "Desired balance computation for [%d] is still not converged after [%d] [%d] iterations",
                         desiredBalanceInput.index(),
-                        i
+                        Duration.ofMillis(currentTime - computationStartedTime).toString(),
```

Comment:
Hmm this makes an ISO8601 duration string like `PT8H6M12.345S` which is _technically_ correct but also kind of unreadable and not consistent with other logged durations. I prefer the following style, although just using the `TimeValue`  string would also be ok here too:

https://github.com/elastic/elasticsearch/blob/97c533a56273c7b1b9269929f5b6f25f6afed0a9/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java#L708-L709